llpc/patch/llpcPatchIntrinsicSimplify.cpp

```diff
@@ -0,0 +1,413 @@
+/*
+ ***********************************************************************************************************************
+ *
+ *  Copyright (c) 2017-2020 Advanced Micro Devices, Inc. All Rights Reserved.
+ *
+ *  Permission is hereby granted, free of charge, to any person obtaining a copy
+ *  of this software and associated documentation files (the "Software"), to deal
+ *  in the Software without restriction, including without limitation the rights
+ *  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ *  copies of the Software, and to permit persons to whom the Software is
+ *  furnished to do so, subject to the following conditions:
+ *
+ *  The above copyright notice and this permission notice shall be included in all
+ *  copies or substantial portions of the Software.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *  SOFTWARE.
+ *
+ **********************************************************************************************************************/
+/**
+ ***********************************************************************************************************************
+ * @file  llpcPatchIntrinsicSimplify.cpp
+ * @brief LLPC source file: contains implementation of class Llpc::PatchIntrinsicSimplify.
+ ***********************************************************************************************************************
+ */
+#define DEBUG_TYPE "llpc-patch-intrinsic-simplify"
+
+#if defined(_MSC_VER) || defined(__MINGW32__)
+// According to Microsoft, one must set _USE_MATH_DEFINES in order to get M_PI
+// from the Visual C++ cmath / math.h headers:
+// https://docs.microsoft.com/en-us/cpp/c-runtime-library/math-constants?view=vs-2019
+#define _USE_MATH_DEFINES
+#endif
+
+#include <math.h>
+
+#include "llvm/InitializePasses.h"
+#include "llvm/Analysis/ScalarEvolution.h"
+#include "llvm/IR/Intrinsics.h"
+#include "llvm/IR/IntrinsicsAMDGPU.h"
+
+#include "llpcPatchIntrinsicSimplify.h"
+#include "llpcPipelineState.h"
+#include "llpcTargetInfo.h"
+
+using namespace llvm;
+using namespace Llpc;
```

Comment:
The code is already wrapped in `namespace Llpc`, why do we need this `using`?

Reply:
@amdrexu What is the recommended style? `using namespace` or wrapping the code in the `.cpp` file into `namespace Llpc`?

Reply:
Seems like the `using namespace` is also used for the `INITIALIZE_PASS_BEGIN` declarations. Removing the wrapping `namespace Llpc` seems to work and removes this duplication.

Reply:
Often, we add LLPC sources files to Llpc namespace like this:

```
<.cpp>
namespace Llpc
{

<Your source codes>

}
```

Reply:
I check builder codes and find Tim uses your style. I think it is OK. You can keep your codes and ignore my comments.